### DISCUSSION TITLE
Ideas/Requests

### DISCUSSION THREAD

**Discussion by x87** [cleolibrary/CLEO-Redux#1] [Ideas]
Hi there!

If you have any idea on what could be implemented as part of CLEO Redux, please share it with me. Would like to see the community's view.

**Comment by DartPower** [comment]
A compatibility & cleo library updates warnings in logs and/or asserts

**Reply by x87** [reply]
> compatibility

CR now runs along with CLEO Library supporting all custom opcodes

**Comment by VitalRus95** [comment]
Hi, Seemann! I'm happy to know you're making CLEO for re3 and reVC! I think it would be interesting to know what CLEO opcodes do you find most likely to work in this new version. As for me, I would say that among the most necessary features are scm-functions, CLEO missions, and external files handling (like .ini). Also I'd like to ask if button emulation opcodes will be restored. From my perspective (though I'm too far from being a pro in this), if there's the standard opcode `00E1` that works fine, then it should be possible to find out which address to use for button emulation. Anyway, I'm looking forward to CLEO Redux updates and thank you for the possibility to make reGTA even more entertaining!

**Reply by x87** [reply]
My goal is not rush things and avoid some poor design decisions that were made earlier. A [notorious example](https://github.com/cleolibrary/CLEO4/issues/38#issuecomment-787465166) is a `chdir` command. Another thing is providing access to potentially harmful APIs should be restricted (see https://github.com/cleolibrary/CLEO4/issues/29). And there are other flaws in the existing CLEO implementation that cause problems to the players (persistent entities saved in the save file creating stray blip markers or permanent pickups, etc).

I have documented those pain-points and I will share my vision with the community to discuss.

As for my personal preference, I want the core implementation to be as tiny and abstract as possible and only provide a few essential commands (start script, end script) and do the rest via the plugins.

**Reply by VitalRus95** [reply]
Plugins sound scary to me. 😃 How will they be written, by the way? I thought plugins also rely on some kind of stable .exe addresses, which is not the case of re project.

**Reply by x87** [reply]
Same as existing CLEO plugins, using a provided SDK. 

Yes, in re3/reVC you can't use static addresses but SDK should provide a translation helper to get the address of a function (similar to 0DD1: get_func_addr_by_cstr_name in CLEO Android).

**Reply by x87** [reply]
>Yes, in re3/reVC you can't use static addresses but SDK should provide a translation helper to get the address of a function (similar to 0DD1: get_func_addr_by_cstr_name in CLEO Android).

It's now available in 0.7.1

**Comment by ShaggEw** [comment]
Are there any plans for CLEO as some kind of module for reVC sources? I mean, to use it and compile it as part of the final exe.
Or maybe some kind of explanation of how to do some stuff independently. For example, scm functions. All other things are not so important when you can compile a custom exe with a custom amount of opcodes and if you work with your custom scm, but scm functions are very mysterious stuff (from a tech perspective), at least for me...

**Reply by x87** [reply]
CLEO Redux is written in Rust. I can't see how it can be made a part of re3/reVC sources. The compiled ASI should be compatible with any version of the game though.

I think I have something better than SCM functions which I'm going to publish very soon. Stay tuned.

**Reply by x87** [reply]
>I think I have something better than SCM functions which I'm going to publish very soon

I meant JavaScript support that is now available. You can run exe functions using Memory class. Also you can obviously use functions in your own JS code.

**Comment by HzanRsxa2959** [comment]
Oh boy, code reloading in real-time. This is gonna be fun.

JK; great work- but please consider the following:
1. An event/callback that runs whenever the script implicitly/explicitly terminates/reloads, so that allocated resources can be deallocated.
2. In fact, events just like [CLEO+](https://github.com/JuniorDjjr/CLEOPlus) in general would be great (doesn't have one for script termination/reload because it's not really needed there, though).
3. A directive/header/command/something like:
**NO_RELOAD**: This script will not be implicitly terminated/reloaded unless a saved game is loaded/a new game is started again.
**NEVER_RELOAD**: This script will not be implicitly terminated/reloaded unless the game is closed and launched again.
_All scripts can still be explicitly terminated/reloaded._
4. Just like **TERMINATE_THIS_CUSTOM_SCRIPT**, a new command: **RELOAD_THIS_CUSTOM_SCRIPT** (self-explanatory).
5. CLEO saves (however you plan on implementing them) should be stored relative to the game's save directory instead of the game's root directory.
6. Maybe we could just do away with opcode numbers in CLEO plugins? While rare, when they do conflict, some bad stuff happens. Examples: MPG.cleo and CLEOImGUI, San Andreas Rampages and CLEO+.
7. An event/callback to run code before the game has loaded like [here](https://wiki.blast.hk/moonloader/lua/onSystemInitialized).
8. A command: **GET_THIS_SCRIPT_DIRECTORY** useful to manage files and folders in script's directory.
9. Although the above is worthwhile only if working with strings is made easier than traditional SCM and even CLEO.

I hope this isn't too much; I suggested all that came to mind.

**Reply by x87** [reply]
@HzanRsxa2959 thanks so much for your feedback and ideas. Let me clarify a few things.


> An event/callback that runs whenever the script implicitly/explicitly terminates/reloads, so that allocated resources can be deallocated.

Agreed

> In fact, events just like CLEO+ in general would be great (doesn't have one for script termination/reload because it's not really needed there, though).

Agreed

> NO_RELOAD: This script will not be implicitly terminated/reloaded unless a saved game is loaded/a new game is started again.

Can you provide an example when this can be useful?

> NEVER_RELOAD: This script will not be implicitly terminated/reloaded unless the game is closed and launched again.

Can you provide an example when this can be useful?


> Just like TERMINATE_THIS_CUSTOM_SCRIPT, a new command: RELOAD_THIS_CUSTOM_SCRIPT (self-explanatory).

I see a benefit in this when coupled with on_reload event, so you can cleanup the resources and re-run the script. Just want to note that it can be implemented with a regular code too (reset all variables and jump to the start of the script)

> CLEO saves (however you plan on implementing them) should be stored relative to the game's save directory instead of the game's root directory.

I don't want to implement CLEO saves. I think they are error-prone and don't help much. Essentially they should've prevented modifying the game state twice and keep adding persistent entities, such as blips or car parkings.  I want to make sure the game does not save anything that is being created by a cleo script. 

If you have a solid use-case when the cleo save is needed, please share, we can discuss.

> Maybe we could just do away with opcode numbers in CLEO plugins? While rare, when they do conflict, some bad stuff happens. Examples: MPG.cleo and CLEOImGUI, San Andreas Rampages and CLEO+.

Agreed, this is one of the flaws in the current CLEO. Pretty much as CLEO vars that use number indexes.

> An event/callback to run code before the game has loaded like here.

Can you provide an example when this can be useful?

> A command: GET_THIS_SCRIPT_DIRECTORY useful to manage files and folders in script's directory.

Noted. This can be done as a plugin.

**Reply by HzanRsxa2959** [reply]
> > NO_RELOAD: This script will not be implicitly terminated/reloaded unless a saved game is loaded/a new game is started again.
>
> Can you provide an example when this can be useful?
>
Mostly when memory has been patched, like in the [IV Pickups](https://github.com/ThirteenAG/III.VC.SA.CLEOScripts/releases/tag/IVPickups) mod, or the [Manual Drive-by Remake](https://github.com/ThirteenAG/III.VC.SA.CLEOScripts/releases/tag/ManualDrivebyRemake) mod.
Although I have yet to see a mod where unloading the script after memory patching crashes the game.
I guess I just want to have more control over the code, because from my experience, unwanted script reloading is a nightmare in MoonLoader. So this option could prevent it. Furthermore, it would make it possible for scripts to have their data persist across saves, i.e., not deallocating and then reallocating resources every time the game is loaded.

> > NEVER_RELOAD: This script will not be implicitly terminated/reloaded unless the game is closed and launched again.
>
> Can you provide an example when this can be useful?
>
Please see the first answer.

> > Just like TERMINATE_THIS_CUSTOM_SCRIPT, a new command: RELOAD_THIS_CUSTOM_SCRIPT (self-explanatory).
>
> I see a benefit in this when coupled with on_reload event, so you can cleanup the resources and re-run the script. Just want to note that it can be implemented with a regular code too (reset all variables and jump to the start of the script)
>
True: Have a deallocation gosub that is called in the on_reload event, as well as calling it whenever the script wants to explicitly reload.

> > CLEO saves (however you plan on implementing them) should be stored relative to the game's save directory instead of the game's root directory.
>
> I don't want to implement CLEO saves. I think they are error-prone and don't help much. Essentially they should've prevented modifying the game state twice and keep adding persistent entities, such as blips or car parkings. I want to make sure the game does not save anything that is being created by a cleo script.
>
> If you have a solid use-case when the cleo save is needed, please share, we can discuss.
>
The fact that entities will not be saved makes a custom save system that much important, in my opinion.
But you shouldn't implement CLEO saves if you don't want to. However, maybe a utility CLEO plugin that helps with creating one's own custom save system would be great.
It's outright impossible in the current CLEO because of the inability to get the CRC32 hash of the current save file before it has been written to disk. And I **really** don't want the save system to depend on save slots.

> > An event/callback to run code before the game has loaded like here.
>
> Can you provide an example when this can be useful?
>
Mostly for patching memory before the game loads, like changing the menu background. I agree that this is trivial.

**Reply by x87** [reply]
>Mostly for patching memory before the game loads, like changing the menu background. I agree that this is trivial.

before the game starts you don't have much data initialized yet, so I think it is only suitable for memory patches. such files would need another file extension to separate them from regular CS scripts. 

I would say it's a use case for an ASI file (load along with the game and patch the memory). But sometimes it can be overkill to fire Visual Studio and compile a full-blown library just to change one address. I will look into what we can do there.

**Reply by HzanRsxa2959** [reply]
Do you know about [Lua Hooker](https://brmodstudio.forumeiros.com/t4641-lua-hooker)? Well, I came across [this](https://brmodstudio.forumeiros.com/t4641p100-lua-hooker#210627) makeHook function that I think is brilliant. This kind of a function in CLEO would be a game-changer; the ability to hook any game function would basically make it possible to create our own events! [Here](https://github.com/Fabio3rs/LuaHooker) is the source code if you are interested.

**Reply by x87** [reply]
this sounds interesting. potentially dangerous too :)

**Reply by HzanRsxa2959** [reply]
> this sounds interesting. potentially dangerous too :)
>
I forgot to consider that security is an important feature of this implementation. I apologize.
But I genuinely do not understand how it could cause a security issue: The code would still be written in SCM which would be managed by CLEO (I assume CLEO would provide security by refusing to run harmful commands); the hook would just manage the time at which such code would be executed.

Though, on the topic of security, is **hex** ... **end** out as well?

**Reply by x87** [reply]
What I mean is that's easy to screw up the game if the hook is implemented incorrectly, which is highly likely as in-game functions vary in calling conventions and input arguments. I haven't heard about Lua Hooker before. Do you know any mods that actually use hooks to some arbitrary methods. What they do?

Hex...end has nothing to do with CLEO, it's a compiler thing. I don't plan to add it in JavaScript as of now.

**Reply by HzanRsxa2959** [reply]
> Do you know any mods that actually use hooks to some arbitrary methods. What they do?
They're rare to find, but off the top of my head:
1. CLEO+: Adds events such as on_save_confirmation, on_char_create, on_vehicle_render, on_object_delete.
2. MoonLoader: Adds events such as onLoadGame, onSaveGame, onStartNewGame.
3. Dynamic Tram Signs?
4. MobileHUD.
5. fastman92 Limit Adjuster: I can't even begin counting the amount of hooks it makes.
But they are made by some of the best scripters in the community, not your everyday mod-maker. And most are ASI rather that SCM/CLEO.
> What I mean is that's easy to screw up the game if the hook is implemented incorrectly, which is highly likely as in-game functions vary in calling conventions and input arguments.
>
True. Giving beginner and even intermediate scripters a hold of such an intricate technique seems like a recipe for disaster.

Thank you, I understand now. I would suggest (though it still depends on how you want it) that this **makeHook** request be dropped.
If a mod truly needs a hook to work, then there is ASI.

**Comment by nathan130200** [comment]
- Add an ability to track when script will terminate. Because usually we allocate some peds, vehicles, blips, etc and there's no way to track when script finish to remove these entities created. And after some time game stop spawning peds, vehicles due too many of them. My suggestion is add an per-script level an event `OnScriptShutdown(isHotReload: boolean)` or `OnScriptCleanup(isHotReload: boolean)` will be a function called before script end or when hot reload, reload this script. So we be able to correctly cleanup stuff. This function will be invoked, even if script was stucked. This will give us more control.

- An way to get `handle` of peds/chars and vehicles. An method such as `getHandle()` for typescript. Because when i need use for example `Char.DoesExist` i must cast to `Char.DoesExist(<any>chaserNpc);`. Idk if this is correct but works.

- Convert all XYZ objects to proper Vec3 objects with relative methods add, subtract, multiply, divide, dist2d, dist3d, etc.

**Comment by x87** [comment]
> Add an ability to track when script will terminate. 

Yes, this would be very helpful. It was requested a few times, however it's difficult to implement which is why it is not there yet. I'm currently focusing on some other projects, so maybe once I get more time and motivation I'll try to implement it.

> An way to get handle of peds/chars and vehicles. 

it can be done by calling .valueOf() method on the char/vehicle instance, or using +operator 

```
let car = Car.Create(...)
let handle = car.valueOf();
// or let handle = +car
```

> Convert all XYZ objects to proper Vec3 objects with relative methods add, subtract, multiply, divide, dist2d, dist3d, etc.

this can be done via third-party libraries. example: https://github.com/VitalRus95/gta-scripts/blob/main/CLEO%20Redux/GTA%20SA/Classes/CVector.ts

**Reply by nathan130200** [reply]
Oh thanks, i didn't noticed the `valueOf` inside the entities. Omg thanks about CVector. I will download these typescript files now to use.

**Reply by nathan130200** [reply]
> Yes, this would be very helpful. It was requested a few times, however it's difficult to implement which is why it is not there yet. I'm currently focusing on some other projects, so maybe once I get more time and motivation I'll try to implement it.

By the way, there's some way to make script don't crash because this? Eg: tracked entities by script (not marked to cleanup/mission entities) crash entire game we do load game/new game unless if markAsNoLongerNeeded() but sometimes inside some complex loops its almost impossible to call everytime this.

**Comment by miloplacas** [comment]
write some for the ios version that is the only one that can play in recent macs

**Comment by Mereuta97** [comment]
> Hi there!
> 
> If you have any idea on what could be implemented as part of CLEO Redux, please share it with me. Would like to see the community's view.

Can support for Epic games can be added on the next version please ? i can't use no mods on gta trilogy because it's on epic games and doesn't have support cleo redux for epic games

**Reply by x87** [reply]
I don't have Epic, only Steam version. Find me on Discord and we can discuss further
